### ISSUE TITLE
webrtc: WASM support

### ISSUE THREAD

**Issue by youngjoon-lee** [libp2p/rust-libp2p#4111]
## Description

Make WebRTC works in WASM environments.

## Motivation

https://github.com/libp2p/rust-libp2p/discussions/3875#discussioncomment-6220525: Discussion for the needs of WebRTC WASM
https://github.com/libp2p/rust-libp2p/pull/4015: WebTransport implementation for WASM which can be a reference of this issue.

## Current Implementation

The current WebRTC transport in rust-libp2p works only for standalone nodes.

## Are you planning to do it yourself in a pull request?

Yes, if there is no plan by libp2p team.

**Comment by thomaseizinger** [comment]
> Yes, if there is no plan by libp2p team.

No plans currently but very grateful for a contribution!

**Comment by mxinden** [comment]
Yes indeed! Very grateful.

**Comment by mxinden** [comment]
Just stumbled upon rust-libp2p `libp2p-wasm-ext` WebRTC bindings.

https://github.com/openmina/rust-libp2p/blob/webrtc/transports/wasm-ext/src/webrtc.js

This is used in https://github.com/openmina/openmina/blob/0e0039bfdf8ae1942b029cf98740129e7847efa8/node/wasm/src/service/libp2p/mod.rs#L71-L82.

Thanks @p-shahi for surfacing this.

@binier since you have implemented the above, mind sharing some details here? Also would you be interested in upstreaming your work or help @youngjoon-lee to do so?

One small note is that we are moving away from the `libp2p-wasm-ext` architecture, towards individual crates using websys. See e.g https://github.com/libp2p/rust-libp2p/pull/4015.

**Comment by youngjoon-lee** [comment]
Awesome. Thank you. I'll take a look at the openmina's implementation and see if I can migrate it into the websys way.

**Comment by binier** [comment]
@mxinden It's been a while since I touched that, so I don't remember all the details, but to speak to the state of the code, I'd say it's WIP and doesn't follow any specs (if there is any for webrtc transport). It works well with go-libp2p (modified), which was the goal. I won't be continuing work on that though, instead we've decided to implement and use custom protocol just based on WebRTC.

@youngjoon-lee Let me know if you'll have any questions, I'll gladly help.
